tests/function/gtVersion.cpp

```diff
@@ -0,0 +1,145 @@
+used,   copied,   reproduced,   modified,   published,   uploaded,  posted,
+transmitted,  distributed, or  disclosed in  any way  without Intel's prior
+express written permission.
+
+No license under any patent, copyright,  trade secret or other intellectual
+property  right  is  granted to  or conferred  upon  you by  disclosure  or
+delivery of the  Materials, either  expressly, by  implication, inducement,
+estoppel or otherwise. Any license  under such intellectual property rights
+must be express and approved by Intel in writing.
+
+--*/
+
+#include <opae/fpga.h>
+
+#include "common_test.h"
+#include "gtest/gtest.h"
+#include "config_int.h"
+
+using namespace common_test;
+using namespace std;
+
+/**
+ * @test       version_01
+ *
+ * @brief      When I retrieve fpga_version information using
+ *             fpgaGetOPAECVersion using a NULL pointer as the return buffer,
+ *             the function returns FPGA_INVALID_PARAM.
+ */
+TEST(LibopaecVersionCommonALL, version_01) {
+	EXPECT_EQ(FPGA_INVALID_PARAM, fpgaGetOPAECVersion(NULL));
+}
+
+/**
+ * @test       version_02
+ *
+ * @brief      When I retrieve fpga_version information using
+ *             fpgaGetOPAECVersion, the returned values match the constants
+ *             defined in config_int.h (and the function returns FPGA_OK)
+ */
+TEST(LibopaecVersionCommonALL, version_02) {
+	fpga_version version = { 0xFF, 0xFF, 0xFFFF };
+
+	EXPECT_EQ(FPGA_OK, fpgaGetOPAECVersion(&version));
+	EXPECT_EQ(INTEL_FPGA_API_VER_MAJOR, version.major);
+	EXPECT_EQ(INTEL_FPGA_API_VER_MINOR, version.minor);
+	EXPECT_EQ(INTEL_FPGA_API_VER_REV, version.patch);
+}
+
+/**
+ * @test       version_03
+ *
+ * @brief      When I retrieve fpga_version information using
+ *             fpgaGetOPAECVersionString using a NULL pointer as the return
+ *             buffer, the function returns FPGA_INVALID_PARAM.
+ */
+TEST(LibopaecVersionCommonALL, version_03) {
+	EXPECT_EQ(FPGA_INVALID_PARAM, fpgaGetOPAECVersionString(NULL, 42));
+}
+
+/**
+ * @test       version_04
+ *
+ * @brief      When I retrieve fpga_version information using
+ *             fpgaGetOPAECVersionString using a size shorter than the minimum
+ *             possible (which is 6) the function returns FPGA_EXCEPTION.
+ */
+TEST(LibopaecVersionCommonALL, version_04) {
+	char str[80];
+
+	EXPECT_EQ(FPGA_EXCEPTION, fpgaGetOPAECVersionString(str, 5));
+}
+
+/**
+ * @test       version_05
+ *
+ * @brief      When I retrieve fpga_version information using
+ *             fpgaGetOPAECVersionString, the returned string represents
+ *             a string comprised of the constants defined in config_int.h
+ *             (and the function returns FPGA_OK)
+ */
+TEST(LibopaecVersionCommonALL, version_05) {
+	char want[80];
+	char have[80];
+
+	snprintf(want, 80, "%d.%d.%d",
+			INTEL_FPGA_API_VER_MAJOR,
+			INTEL_FPGA_API_VER_MINOR,
+			INTEL_FPGA_API_VER_REV
+		);
+
+	EXPECT_EQ(FPGA_OK, fpgaGetOPAECVersionString(have, 80));
+	EXPECT_STREQ(want, have);
```

Comment:
Can we replace `want` with `"INTEL_FPGA_API_VER_MAJOR.INTEL_FPGA_API_VER_MINOR.INTEL_FPGA_API_REV"`

Reply:
That won't work. Preprocessor macros are not expanded within string constants.

If the constants were strings instead of integers, you could do

    INTEL_FPGA_API_VER_MAJOR "." INTEL_FPGA_API_VER_MINOR "." INTEL_FPGA_API_REV

but they aren't, so we can't. ;)

Reply:
aaah... 👍